concepts/code-organization/introduction.md

```diff
@@ -0,0 +1,161 @@
+
+## Include Guards
+
+It does not matter if the same file is included multiple times within a project.
+Header files should not contain definitions.
+The complete project cannot have the same definition more than once. 
+This is called the "One definition rule".
+It will be enforced by the compiler.
+
+It is easy to avoid multiple unintended inclusions of the same definition with include guards.
+They are formed by a special procedure during the compilation stages.
+The aim is to only include a file only if a certain variable has not been set and then set it once the file is included.
+Often the sequence of this variable is chosen as a variation of the name of the file.
+Another method is the generation of a UUID to reduce the risk of accidentally using the sequence twice.
+The syntax can be seen below with `MY_HEADER_FILE_H` as a variable.
+
+```cpp
+#ifndef MY_HEADER_FILE_H /* any name uniquely mapped to file name */
+#define MY_HEADER_FILE_H
+// file content
+#endif
+```
+A modern form, that will achieve the same result is the #pragma directive `#pragma once`.
+The problem is, that pragmas are not an official part of the C++ language and the implementation differs from compiler to compiler.
+
+## Namespaces
+
+An important method for code organization are namespaces.
+Two functions `int foo()` can be used side-by-side if they are defined in different namespaces.
+They can be called with the scope resolution operator `::`.
+Namespaces can be nested.
+They are kept as they are after includes.
+
+```cpp
+namespace my_ns {
+    int foo() {
+        return 44;
+    }
+    namespace my_inner_ns {
+        int baz() {
+            return 90;
+        }
+    }
+}
+namespace my_other_ns {
+    int foo() {
+        return -2;
+    }
+}
+
+int myresult{my_ns::foo() + my_other_ns::foo() * my_ns::my_inner_ns::baz()};
+```
+
+Deeply nested namespaces might be too verbose.
+It is possible to remove the verbosity with a namespace import via `using`.
+To keep some differentiation aliases might be useful.
+
+```cpp
+int my_other_result{my_ns::my_inner_ns::baz()};
+
+using namespace my_ns::my_inner_ns; // importing the complete namespace
+int also_other_result{baz()};
+
+namespace m = my_ns; // setting an alias
+namespace o = my_other_ns;
+
+int also_my_result{m::foo() + o::foo() * m::my_inner_ns::baz()};
+```
+
+## Standard Library Header Files
+ 
+The standard library offers many common functions, algorithms, and data structures.
+They are grouped into header files.
+They are included with angled braces `<>` instead of double quotes `"`.
+The difference with the braces is the location, where the compiler searches for the header files.
+The search in the current project is skipped for the angled braces version, and it directly starts in the system's include directories.
+The standard library headers are located in the `std` namespace.
```

Comment:
the old C std lib is also available both in namespace std or globally
`#include <stdio.h>`
vs
`#include <cstdio>`

Reply:
I would not talk too extensively about the finer differences. If the student understands this chapter well it should be easier to look up what will change when they use .h std files. Right?